flink-runtime/src/test/java/org/apache/flink/runtime/operators/coordination/CoordinatorEventsExactlyOnceITCase.java

```diff
@@ -285,41 +348,130 @@ public void handleEventFromOperator(int subtask, OperatorEvent event) throws Exc
                         String.format("Don't recognize event '%s' from task %d.", event, subtask));
             }
 
-            if (periodicTask != null) {
-                throw new Exception("periodic already running");
+            synchronized (recoveredTaskRunning) {
+                // signal the previous task that its recovered task is now running
+                final CompletableFuture<?> prevTaskFuture = recoveredTaskRunning.peekLast();
+                if (prevTaskFuture != null) {
+                    prevTaskFuture.complete(null);
+                }
+                // add a future for this task
+                recoveredTaskRunning.addLast(new CompletableFuture<>());
             }
-            periodicTask =
-                    executor.scheduleWithFixedDelay(this, delay, delay, TimeUnit.MILLISECONDS);
+
+            // first, we hand this over to the mailbox thread, so we preserve order on operations,
+            // even if the action is only to do a thread safe scheduling into the scheduledExecutor
+            runInMailbox(
+                    () -> {
+                        checkState(!workLoopRunning);
+                        checkState(subtaskGateway != null);
+
+                        workLoopRunning = true;
+                        scheduleSingleAction();
+                    });
         }
 
         @Override
         public void subtaskFailed(int subtask, @Nullable Throwable reason) {
```

Comment:
Actually, there is no true subtask failure in this test case. This `subtaskFailed` is caused by global failure. I think we should draw some unrecognized events to fail subtask.

Reply:
Could you clarify what you mean by "draw some unrecognized events to fail subtask"?

Reply:
I means sending `UnrecognizedEvent` to operator to fail running task. Currently, running tasks are not failed due to themselves but global failover.

Reply:
I think that is a nice idea. Ideally we have both (globalFailover from master, and task failure).
I expect this should just work, as it doesn't matter to the scheduler where the failure comes from.

Let's apply that as a follow-up patch, that we have both types of failures.

Reply:
It should require some minor changes to `subtaskReset`, `notifyCheckpointComplete` and `notifyCheckpointAborted`. Personally, I think `SubtaskGateway` is more vital to subtask failure than job failure. For solely job failure, recreating a brand new coordinator on `resetToCheckpoint` should pass this test also.

Reply:
To my understanding, the problem is independent of that. It is caused by the fact that a new subtask is scheduled but the Context stays the same. Even if the Coordinator is recreated, as long as the Context is not re-created, the problem exists. And in some sense that is what is happening now: we re-create that part of the context that is responsible for event sending.

Reply:
Agree. For me, the important part is that `RecreateOnResetOperatorCoordinator` bring an opportunity to  do clean blocking termination of old coordinator. So, even the context are not re-created, but there is no sending activities from old coordinator.